### ISSUE TITLE
esbuild support

### ISSUE THREAD

**Issue by abrvsk** [nestjs/nest-cli#731]
## Feature Request
Add support for esbuild as the build system.

## Is your feature request related to a problem? Please describe.
Building modules on large project takes sizeable amount of time, which makes debugging a bit slow. esbuild can potentially increase the build speed up to 86 times.

## Describe the solution you'd like
Include esbuild into the list of supported build systems.

## Teachability, Documentation, Adoption, Migration Strategy
[esbuild repo](https://github.com/evanw/esbuild)

## What is the motivation / use case for changing the behavior?
Speed. Speed. Speed.

**Comment by jmcdo29** [comment]
This looks pretty impressive, but a major problem is that it does not support decorators, which would kind of go against how Nest does all of its DI. I'm definitely going to keep an eye on the library though. @kamilmysliwiec if this ever gets to a point where it _could_ support Nest, what would you think?

**Comment by kamilmysliwiec** [comment]
> Separate from that, decorators in TypeScript have been around for a while behind the experimentalDecorators flag, and have found a lot of adoption in the web development community. I think esbuild would be useful to many more people if it had support for TypeScript-style decorators. So I'm planning to implement those at some point.

https://github.com/evanw/esbuild/issues/104#issuecomment-627645920

It seems that TS decorators might be supported at some point. Let's circle back when it's implemented. For now, `esbuild` isn't compatible with Nest

**Comment by aelbore** [comment]
> This looks pretty impressive, but a major problem is that it does not support decorators, which would kind of go against how Nest does all of its DI. I'm definitely going to keep an eye on the library though. @kamilmysliwiec if this ever gets to a point where it _could_ support Nest, what would you think?

have you tried @swc/core? they supports decorators
https://swc-project.github.io

**Comment by techvlad** [comment]
> > This looks pretty impressive, but a major problem is that it does not support decorators, which would kind of go against how Nest does all of its DI. I'm definitely going to keep an eye on the library though. @kamilmysliwiec if this ever gets to a point where it _could_ support Nest, what would you think?
> 
> have you tried @swc/core? they support decorators
> https://swc-project.github.io

SWC has an issue with decorators
https://github.com/swc-project/swc/issues/666

**Comment by tooolbox** [comment]
> An initial implementation of TypeScript decorators has been released in version 0.4.10. You can read more about this in the release notes. Please let me know if you encounter any issues.

https://github.com/evanw/esbuild/issues/104#issuecomment-640559073

**Comment by jmcdo29** [comment]
> > An initial implementation of TypeScript decorators has been released in version 0.4.10. You can read more about this in the release notes. Please let me know if you encounter any issues.
> 
> [evanw/esbuild#104 (comment)](https://github.com/evanw/esbuild/issues/104#issuecomment-640559073)

This still doesn't support `emitDecoratorMetadata`, which Nest makes use of, so I don't think it can be used yet. Also, it doesn't support commonjs unless working with bundles which also becomes a problem due to Nest's optional require statements

**Comment by dethereum** [comment]
I think swc does work with nestjs now. Using the @swc/cli @swc/core packages and this .swcrc configuration

```
{
    "jsc": {
      "loose": true,
      "target": "es2020",
      "parser": {
        "syntax": "typescript",
        "decorators": true
      },
      "transform": {
        "legacyDecorator": true,
        "decoratorMetadata": true
      }
    },
    "module": {
        "type": "commonjs",
        "strict": true,
        "strictMode": true,
        "lazy": true,
        "noInterop": true
    }
  }
  ```
  
  I run these commands and everything works fine. See this [PR](https://github.com/swc-project/swc/pull/939)

```
 "build": "swc src -s -d ./dist",
 "start": "pnpm build && node dist/main.js",
 ```

**Comment by kamilmysliwiec** [comment]
Using either `swc` and `esbuild` with Nest should be doable. The biggest issue with these compilers is a lack of type-checking which is one of the major benefits of TS. We currently don't plan to swap `tsc` with either one, but you should be able to safely use them in your projects.

**Comment by axe-me** [comment]
Can we use `swc` or `esbuild` in dev server only? but still, use `tsc` in production build? 
In dev, we can rely on IDEs to do type checks. This will gain a lot DX from fast rebuild.

**Comment by jmcdo29** [comment]
Sure, just depends on the scripts you set up and where you run the production build. You can absolutely use `swc` or `esbuild` in dev (though I think `esbuild` still doesn't emit metadata) and let typescript be your CI/CD compiler. Scripts lead to versatility, right? All about customization.

**Comment by wxs77577** [comment]
@dethereum Could you please tell me how to use  `swc` with nestjs monorepo?

**Comment by xiaoyu-tamu** [comment]
Anyone made swc work with nestjs monorepo?

you can also disable typecheck during development as workaround to speed up compile 

```ts
// webpack.config.js
const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin');

module.exports = options => {
  if (process.env.NODE_ENV === 'production') return options;

  return {
    ...options,
    plugins: options.plugins.filter(x => !(x instanceof ForkTsCheckerWebpackPlugin)),
  };
};

```

![image](https://user-images.githubusercontent.com/33362998/110224719-b5cfde00-7ea3-11eb-9244-b51adf588b64.png)

**Comment by techvlad** [comment]
For those who find this Issue, I created an example with `esbuild` and `swc` for local development
https://github.com/techvlad/nestjs-build
@axe-me @abrvsk

**Comment by axe-me** [comment]
@techvlad that esbuild plugin you use is using tsc to compile ts file with decorators.

**Comment by techvlad** [comment]
@axe-me I'm using https://github.com/anatine/esbuildnx/tree/main/packages/esbuild-decorators plugin, this plugin doesn't use `tsc`

**Comment by peterfication** [comment]
I tried to get the solution of @techvlad  running with TypeORM but had no luck unfortunately.

See https://github.com/techvlad/nestjs-build/pull/2

**Comment by axe-me** [comment]
@techvlad this one clearly using tsc to compile ts file with decorator inside, see: https://github.com/anatine/esbuildnx/blob/main/packages/esbuild-decorators/src/lib/esbuild-decorators.ts#L11

I just made a vitejs plugin to run my dev server with super fast HMR. it supports Nestjs out of box. However you have to set the tsCompiler to swc in order to make decorators works.
Repo: https://github.com/axe-me/vite-plugin-node

**Comment by nartc** [comment]
Hi all, I'm late for the party here.

I've been working on `swc` support for Nx, specifically Node, Express, and NestJS. And `swc` works pretty well with most common use-cases of NestJS (decorators and such...). Things that don't work are Transformers (`nestjs/swagger/plugin` etc...) because currently `swc` doesn't provide a way to hook into the compilation process.

For `webpack`, `swc` has a loader (https://swc.rs/docs/usage-swc-loader) to use with `webpack`, it's pretty much plug and play. People can still benefit from type-check with `ForkTsCheckerWebpackPlugin`.

**Comment by tonivj5** [comment]
@nartc out of curiosity, only the decorators of `class-transformer/validator` are broken? 😆

**Comment by nartc** [comment]
> @nartc out of curiosity, only the decorators of `class-transformer/validator` are broken? 😆

Well, the Controller and stuff from NestJS works lol. My automapper's decorator also works.

**Comment by xiaoyu-tamu** [comment]
> @nartc out of curiosity, only the decorators of `class-transformer/validator` are broken? 😆

my workaround is always provide a default value `undefined as any`

```ts
export class CreateUserDto {
  @IsEmail()
  email: string = undefined as any;
}
```

**Comment by tonivj5** [comment]
The problem is related to optional properties `?` and the code generated around them. I was wondering if it only happened with `class-transformer/validator` (who knows 😆) or it's general...

**Comment by yharaskrik** [comment]
> > @nartc out of curiosity, only the decorators of `class-transformer/validator` are broken? 😆
> 
> my workaround is always provide a default value `undefined as any`
> 
> ```ts
> export class CreateUserDto {
>   @IsEmail()
>   email: string = undefined as any;
> }
> ```

I can confirm that assigning a default value to each prop does indeed make it work (obviously this is less than ideal but it is something). `class-validator` seems to work regardless but `class-transformer` specifically `plainToClass` needs a default value. As @nartc mentioned this is also an issue with `swc`

**Comment by RIP21** [comment]
Another option is to fix it on the `class-transformer` side is by using `patch-package` 
https://github.com/ds300/patch-package
And apply this fix using it.
https://github.com/typestack/class-transformer/pull/1007

Here is the patch I've got at the end:
```
diff --git a/node_modules/class-transformer/cjs/TransformOperationExecutor.js b/node_modules/class-transformer/cjs/TransformOperationExecutor.js
index 81aec2a..9c1e13b 100644
--- a/node_modules/class-transformer/cjs/TransformOperationExecutor.js
+++ b/node_modules/class-transformer/cjs/TransformOperationExecutor.js
@@ -251,7 +251,7 @@ class TransformOperationExecutor {
                     if ((this.transformationType === enums_1.TransformationType.PLAIN_TO_CLASS ||
                         this.transformationType === enums_1.TransformationType.CLASS_TO_CLASS) &&
                         // eslint-disable-next-line @typescript-eslint/unbound-method
-                        ((descriptor && !descriptor.set) || newValue[newValueKey] instanceof Function))
+                        ((descriptor && !descriptor.set && !('value' in descriptor)) || newValue[newValueKey] instanceof Function))
                         //  || TransformationType === TransformationType.CLASS_TO_CLASS
                         continue;
                 }
diff --git a/node_modules/class-transformer/esm2015/TransformOperationExecutor.js b/node_modules/class-transformer/esm2015/TransformOperationExecutor.js
index 0ce77a6..0fb863c 100644
--- a/node_modules/class-transformer/esm2015/TransformOperationExecutor.js
+++ b/node_modules/class-transformer/esm2015/TransformOperationExecutor.js
@@ -248,7 +248,7 @@ export class TransformOperationExecutor {
                     if ((this.transformationType === TransformationType.PLAIN_TO_CLASS ||
                         this.transformationType === TransformationType.CLASS_TO_CLASS) &&
                         // eslint-disable-next-line @typescript-eslint/unbound-method
-                        ((descriptor && !descriptor.set) || newValue[newValueKey] instanceof Function))
+                        ((descriptor && !descriptor.set && !('value' in descriptor)) || newValue[newValueKey] instanceof Function))
                         //  || TransformationType === TransformationType.CLASS_TO_CLASS
                         continue;
                 }
diff --git a/node_modules/class-transformer/esm5/TransformOperationExecutor.js b/node_modules/class-transformer/esm5/TransformOperationExecutor.js
index a92e6d5..7283f0c 100644
--- a/node_modules/class-transformer/esm5/TransformOperationExecutor.js
+++ b/node_modules/class-transformer/esm5/TransformOperationExecutor.js
@@ -251,7 +251,7 @@ var TransformOperationExecutor = /** @class */ (function () {
                     if ((this_1.transformationType === TransformationType.PLAIN_TO_CLASS ||
                         this_1.transformationType === TransformationType.CLASS_TO_CLASS) &&
                         // eslint-disable-next-line @typescript-eslint/unbound-method
-                        ((descriptor && !descriptor.set) || newValue[newValueKey] instanceof Function))
+                        ((descriptor && !descriptor.set && !('value' in descriptor)) || newValue[newValueKey] instanceof Function))
                         return "continue";
                 }
                 if (!this_1.options.enableCircularCheck || !this_1.isCircular(subValue)) {

```

And surely get subscribed to https://github.com/swc-project/swc/issues/2117 to know when this is fixed from `swc` side as I guess `tsc` output is considered a single source of truth.

**Comment by tonivj5** [comment]
> for swc users, this might be issue if you're using `class-transform`
> 
> [swc-project/swc#2117](https://github.com/swc-project/swc/issues/2117)

This has been fixed by swc https://github.com/swc-project/swc/pull/3459 🚀 Waiting for the next release 🎉

**Comment by pharindoko** [comment]
It should be possible to use decorators with esbuild, right ?
https://github.com/evanw/esbuild/issues/104#issuecomment-1028686596
is this still the showstopper ?

**Comment by jmcdo29** [comment]
![image](https://user-images.githubusercontent.com/28268680/154290153-1b19d99c-173b-4229-b69c-bc8e06601547.png)

The `emitDecoratorMetadata` flag **must** be supported for Nest to work properly.

**Comment by pharindoko** [comment]
You're damn right :-/ @jmcdo29

**Comment by thomaschaaf** [comment]
But you can just compile using the plugin I built. It will use tsc for the files which contain decorators and esbuild for everything else. It's still A LOT quicker. (https://github.com/thomaschaaf/esbuild-plugin-tsc)

**Comment by kamilmysliwiec** [comment]
@thomaschaaf I believe 90% of files in Nest codebases _will contain_ decorators. Could you share some benchmarks? 🙌

**Comment by xiaoyu-tamu** [comment]
I’m curious why not just use sec if you want speed up? What’s the benefit choose esbuild?

**Comment by manuschillerdev** [comment]
I am trying to evaluate the possibilites for using swc with nestjs. Currently, it looks very promising, since swc support `emitDecoratorMetaData` now :) 

https://github.com/manuschillerdev/nestjs-swc
Try `npm run start:dev` for an experimental dev server.

Since I am not the most experience nestjs developer - what would be the edge cases to test here?

TLDR;
* Tests seems to work without any problems with `@swc/jest` 🎉
* The app seems to work great as well in dev - typeorm sometimes loses the connection, though.
* Currently I have no Idea how I would use swc to bundle the app :) 

Might write something on dev.to about the experiment, if someone is interested?

**Comment by kamilmysliwiec** [comment]
Thanks @manuschillerdev!

I think the most substantial caveat is that CLI plugins (`@nestjs/swagger` and `@nestjs/graphql`) won't work as expected with SWC as they require tsc AST transformations to be applied.

**Comment by manuschillerdev** [comment]
ah I see - thanks @kamilmysliwiec 
swc has several compat approaches (e.g. https://crates.io/crates/swc_estree_compat), but I'm sure it would be a massive effort to either produce a valid TSC Ast by swc, oder rewrite your mentioned plugins to use another AST format :/

**Comment by Karibash** [comment]
When build with swc, there seems to be a problem with DI for property decorators.
https://github.com/swc-project/swc/issues/3592

**Comment by jackel119** [comment]
I've made this issue #1549 in order to track SWC support specifically, since ESBuild looks like a dead end in the medium term due to not supporting `emitDecoratorMetadata`.

**Comment by yharaskrik** [comment]
> I've made this issue #1549 in order to track SWC support specifically, since ESBuild looks like a dead end in the medium term due to not supporting `emitDecoratorMetadata`.

I have been using SWC to compile Nest recently and it seems to be working (all my code is managed in an Nx monorepo and we are using the @nrwl/js plugin with the swc compiler for buildable libraries)

**Comment by jackel119** [comment]
@yharaskrik are you using CLI plugins like swagger?

**Comment by yharaskrik** [comment]
> @yharaskrik are you using CLI plugins like swagger? 

Ah, I'm actually only using it to compile e2e tests. Our deployed code with swagger is still using the regular builders no swc. We are only using swc to compile plain node libs right now in deployed code.

**Comment by Nickersoft** [comment]
I'd just like to chime in here and say that I have been using `@swc/cli` to compile my production NestJS repo for months now and it seems to work perfectly. I use `@nestjs/graphql` extensively, as my app is primarily a GraphQL server. I run the build in tandem with `tsc --noEmit` using [concurrently](https://npmjs.org/package/concurrently). 

Here is my `.swcrc`:

```json
{
  "module": {
    "type": "es6"
  },
  "jsc": {
    "baseUrl": ".",
    "target": "es2020",
    "parser": {
      "syntax": "typescript",
      "decorators": true,
      "importMeta": true
    },
    "transform": {
      "legacyDecorator": true,
      "decoratorMetadata": true
    },
    "keepClassNames": true,
  }
}
```

I mostly was forced to go with a solution like `swc` because some of my dependencies are ESM-only, which Webpack doesn't have great support for.

**Comment by sebastiangug** [comment]
@Nickersoft  did you find a way to be able to use the CLI plugin? 

I've been using vite + swc pretty alright but the right way of enabling that functionality has eluded me

**Comment by gerardolima** [comment]
esbuild [version 0.18.5](https://github.com/evanw/esbuild/releases/tag/v0.18.5) was just released. This version [supports decorators](https://github.com/evanw/esbuild/issues/104#issuecomment-1597922979). Is support to building with esbuild still being considered?

**Comment by jmcdo29** [comment]
@gerardolima This looks to just be for the new JS decorator syntax, which Nest doesn't use. WE still use "legacy" decorators from Typescript because we need the emitted metadata that Typescript providers. If you're looking for a faster compilation cycle, Nest supports SWC in version 10 directly with the `-b swc` flag to the CLI
